pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +276,176 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
+func ValidateCSINodeUpdate(new, old *storage.CSINode) field.ErrorList {
+	allErrs := ValidateCSINode(new)
+
+	// Spec is read-only
+	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
+	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "field is immutable"))
+	}
+	return allErrs
+}
+
+// ValidateCSINodeSpec tests that the specified CSINodeSpec has valid data.
+func validateCSINodeSpec(
+	spec *storage.CSINodeSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateCSINodeDrivers(spec.Drivers, fldPath.Child("drivers"))...)
+	return allErrs
+}
+
+// ValidateCSINodeDrivers tests that the specified CSINodeDrivers have valid data.
+func validateCSINodeDrivers(drivers []storage.CSINodeDriver, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	driverNamesInSpecs := make(sets.String)
+	for i, driver := range drivers {
+		idxPath := fldPath.Index(i)
+		allErrs = append(allErrs, validateCSINodeDriver(driver, driverNamesInSpecs, idxPath)...)
+	}
+
+	return allErrs
+}
+
+// validateCSINodeDriverNodeID tests if Name in CSINodeDriver is a valid node id.
+func validateCSINodeDriverNodeID(nodeID string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// It is possible that nodeID will be empty on Kubelet startup, and is populated
+	// later when/if the Driver is registered.
+	if len(nodeID) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, nodeID))
+	}
+	if len(nodeID) > 128 {
+		allErrs = append(allErrs, field.Invalid(fldPath, nodeID, "nodeID must be 128 characters or less"))
+	}
+	return allErrs
+}
+
+// validateCSINodeDriver tests if CSINodeDriver has valid entries
+func validateCSINodeDriver(driver storage.CSINodeDriver, driverNamesInSpecs sets.String, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, apivalidation.ValidateCSIDriverName(driver.Name, fldPath.Child("name"))...)
+	allErrs = append(allErrs, validateCSINodeDriverNodeID(driver.NodeID, fldPath.Child("nodeID"))...)
+
+	// check for duplicate entries for the same driver in specs
+	if driverNamesInSpecs.Has(driver.Name) {
+		allErrs = append(allErrs, field.Duplicate(fldPath.Child("name"), driver.Name))
+	}
+	driverNamesInSpecs.Insert(driver.Name)
+	topoKeys := make(sets.String)
+	for _, key := range driver.TopologyKeys {
+		if len(key) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath, key))
+		}
+
+		if topoKeys.Has(key) {
+			allErrs = append(allErrs, field.Duplicate(fldPath, key))
+		}
+		topoKeys.Insert(key)
+
+		for _, msg := range isValidTopologyKey(key) {
+			allErrs = append(allErrs, field.Invalid(fldPath, driver.TopologyKeys, msg))
+		}
+	}
+
+	return allErrs
+}
+
+func isValidTopologyKey(key string) []string {
+	var errs []string
+	parts := strings.Split(key, "/")
```

Comment:
can we replace most of this with [ValidateLabelName](https://sourcegraph.com/github.com/kubernetes/apimachinery@master/-/blob/pkg/apis/meta/v1/validation/validation.go#L58)?

Reply:
Prefix is different from name.  We can only use ValidateLabelName for the name, not the prefix.

Reply:
doesn't `validation.IsQualifiedName()` cover this?

Reply:
Indeed!  Replaced isValidTopologyKey with validation.IsQualifiedName().